### DISCUSSION TITLE
How to response calculated data with dependency

### DISCUSSION THREAD

**Discussion by sytranvn** [fastapi/fastapi#12905] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from minio import Minio
from typing import Annotated

app = FastAPI()

def get_minio_client():
    return Minio(
        "localhost:9000",
        access_key="MINIO_KEY",
        secret_key="MINIO_SECRET",
        secure=settings.ENVIRONMENT != "local"
    )


MinioDep = Annotated[Minio, Depends(get_minio_client)]


class MyModel(SQLModel):
  id: uuid.UUID
  resource: str

class MyModelPublic(MyModel):
  """
  presigned url using minio.get_presigned_url(
            "GET",
            mybucket,
            resource,
            expires=timedelta(days=1)
        )
  """
  resource: str 


@app.get("/{id}", minio: MinioDep, response_model=MyModelPublic)
def read_root():
    mymodel = MyModel(id=uuid.uuid4(), resouce="file.mp3")
    return mymodel
```

### Description

My model has a `resouce` field that holds object name in a storage bucket. But on response, I want to send the presigned url instead. And I want to apply that to all endpoints that depend on the model.

### Operating System

Linux

### FastAPI Version

0.115.4

### Pydantic Version

2.9.2

### Python Version

3.12.6

**Comment by jgould22** [comment]
You may be able to use pydantic's computed fields for this 

https://docs.pydantic.dev/2.0/usage/computed_fields/

**Reply by sytranvn** [reply]
I have read this document. But the `minio` client instance is not part of the model.

**Reply by jgould22** [reply]
Ah looks like you are trying to do the same as this person 

https://github.com/pydantic/pydantic/discussions/6415

**Reply by sytranvn** [reply]
Yeah, this almost helps if computed data is just more static. 

```py

class MyModelPublic(MyModel):
  """
  presigned url using 
  minio.get_presigned_url(
      "GET",
      mybucket,
      resource,
      expires=timedelta(days=1)
  )
  """
  resource: str
  @field_serializer('resource')
  def presign_url(self, v: str, info: SerializationInfo):
    if not v:
        return v
    context = info.context
    if context:
        minio = context.get("minio")
        mybucket = context.get("bucket")
    else:
      minio=get_minio_client()
      mybucket = "hola"
    return minio.get_presigned_url(
        "GET",
        mybucket,
        v,
        expires=timedelta(days=1)
    )
```

I think it's better to feed the minio instance into the context of reponse model. Because initializing new minio instance every time the field_serializer called is bad.

**Reply by sytranvn** [reply, marked as answer]
I updated the code and I think this enough for my use.

```py
from datetime import timedelta
import logging
from fastapi import Depends, FastAPI
from minio import Minio
from typing import Annotated
from sqlmodel import SQLModel
from pydantic import SerializationInfo, field_serializer
import uuid

app = FastAPI()

def get_minio_client():
    return Minio(
        "localhost:9000",
        access_key="KEY",
        secret_key="SECRET",
        secure=False
    )


MinioDep = Annotated[Minio, Depends(get_minio_client)]


class MyModel(SQLModel):
  id: uuid.UUID
  resource: str


context = None

def get_context():
  logging.warning("Called get_context")
  return dict(
    minio = get_minio_client(),
    bucket = "hola"
  )

class MyModelPublic(MyModel):

  """
  presigned url using 
  minio.get_presigned_url(
      "GET",
      mybucket,
      resource,
      expires=timedelta(days=1)
  )
  """
  resource: str
  @field_serializer('resource')
  def presign_url(self, v: str, _info: SerializationInfo):
    if not v:
        return v
    global context
    if context is None:
      context = get_context()
    minio = context["minio"]
    mybucket = context["bucket"]
    return minio.get_presigned_url(
        "GET",
        mybucket,
        v,
        expires=timedelta(days=1)
    )


@app.get("/{id}", response_model=MyModelPublic)
def read_root(id: str):
    mymodel = MyModel(id=uuid.uuid4(), resource="file.mp3")
    return mymodel
```
